[CI] DataNodeRequestSenderTests testLimitConcurrentNodes failed

### CI Link

https://gradle-enterprise.elastic.co/s/u5pu6chk3lu72/tests/task/:x-pack:plugin:esql:test/details/org.elasticsearch.xpack.esql.plugin.DataNodeRequestSenderTests/testLimitConcurrentNodes?top-execution=1

### Repro line

`./gradlew ":x-pack:plugin:esql:test" --tests "org.elasticsearch.xpack.esql.plugin.DataNodeRequestSenderTests.testLimitConcurrentNodes" -Dtests.seed=1053EA2845BB1589 -Dtests.locale=ca-AD -Dtests.timezone=Europe/Luxembourg -Druntime.java=25	`

### Does it reproduce?

No

### Applicable branches

main

### Failure history

_No response_

### Failure excerpt

    java.lang.AssertionError: 
    Expected: <2>
         but: was <1>

--- a/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/plugin/DataNodeRequestSenderTests.java
+++ b/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/plugin/DataNodeRequestSenderTests.java
@@ -73,6 +73,8 @@ import static org.hamcrest.Matchers.empty;
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.hasSize;
 import static org.hamcrest.Matchers.in;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.lessThanOrEqualTo;
 import static org.hamcrest.Matchers.not;
 
 public class DataNodeRequestSenderTests extends ComputeTestCase {
@@ -344,7 +346,7 @@ public class DataNodeRequestSenderTests extends ComputeTestCase {
             });
         }));
         assertThat(sent.size(), equalTo(shards));
-        assertThat(maxConcurrentRequests.get(), equalTo(concurrency));
+        assertThat(maxConcurrentRequests.get(), is(lessThanOrEqualTo(concurrency)));
         assertThat(response.totalShards, equalTo(shards));
         assertThat(response.successfulShards, equalTo(shards));
         assertThat(response.failedShards, equalTo(0));

--- a/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/plugin/DataNodeRequestSenderTests.java
+++ b/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/plugin/DataNodeRequestSenderTests.java
@@ -352,6 +352,53 @@ public class DataNodeRequestSenderTests extends ComputeTestCase {
         assertThat(response.failedShards, equalTo(0));
     }
 
+    public void testLimitConcurrentNodesForced() {
+        final int shards = 10;
+        var targetShards = new ArrayList<DataNodeRequestSender.TargetShard>(shards);
+        for (int i = 0; i < shards; i++) {
+            targetShards.add(
+                targetShard(
+                    new DataNodeRequest.Shard(new ShardId("index", "n/a", i), SplitShardCountSummary.UNSET),
+                    DiscoveryNodeUtils.builder("node-" + i).roles(Set.of(DATA_HOT_NODE_ROLE)).build()
+                )
+            );
+        }
+
+        var concurrency = randomIntBetween(1, 2);
+        AtomicInteger maxConcurrentRequests = new AtomicInteger(0);
+        AtomicInteger concurrentRequests = new AtomicInteger(0);
+        var sent = ConcurrentCollections.<NodeRequest>newQueue();
+        var response = safeGet(sendRequests(randomBoolean(), concurrency, targetShards, (node, shardIds, aliasFilters, listener) -> {
+            concurrentRequests.incrementAndGet();
+
+            while (true) {
+                var priorMax = maxConcurrentRequests.get();
+                var newMax = Math.max(priorMax, concurrentRequests.get());
+                if (newMax <= priorMax || maxConcurrentRequests.compareAndSet(priorMax, newMax)) {
+                    break;
+                }
+            }
+
+            sent.add(nodeRequest(node, shardIds));
+            runWithDelay(() -> {
+                while (maxConcurrentRequests.get() < concurrency) {
+                    try {
+                        Thread.sleep(100);
+                    } catch (InterruptedException e) {
+                        throw new RuntimeException(e);
+                    }
+                }
+                concurrentRequests.decrementAndGet();
+                listener.onResponse(new DataNodeComputeResponse(DriverCompletionInfo.EMPTY, Map.of()));
+            });
+        }));
+        assertThat(sent.size(), equalTo(shards));
+        assertThat(maxConcurrentRequests.get(), equalTo(concurrency));
+        assertThat(response.totalShards, equalTo(shards));
+        assertThat(response.successfulShards, equalTo(shards));
+        assertThat(response.failedShards, equalTo(0));
+    }
+
     public void testSkipNodes() {
         var targetShards = List.of(
             targetShard(shard1, node1),

--- a/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/plugin/DataNodeRequestSenderTests.java
+++ b/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/plugin/DataNodeRequestSenderTests.java
@@ -324,45 +324,8 @@ public class DataNodeRequestSenderTests extends ComputeTestCase {
             );
         }
 
-        var concurrency = randomIntBetween(1, 2);
-        AtomicInteger maxConcurrentRequests = new AtomicInteger(0);
-        AtomicInteger concurrentRequests = new AtomicInteger(0);
-        var sent = ConcurrentCollections.<NodeRequest>newQueue();
-        var response = safeGet(sendRequests(randomBoolean(), concurrency, targetShards, (node, shardIds, aliasFilters, listener) -> {
-            concurrentRequests.incrementAndGet();
-
-            while (true) {
-                var priorMax = maxConcurrentRequests.get();
-                var newMax = Math.max(priorMax, concurrentRequests.get());
-                if (newMax <= priorMax || maxConcurrentRequests.compareAndSet(priorMax, newMax)) {
-                    break;
-                }
-            }
-
-            sent.add(nodeRequest(node, shardIds));
-            runWithDelay(() -> {
-                concurrentRequests.decrementAndGet();
-                listener.onResponse(new DataNodeComputeResponse(DriverCompletionInfo.EMPTY, Map.of()));
-            });
-        }));
-        assertThat(sent.size(), equalTo(shards));
-        assertThat(maxConcurrentRequests.get(), is(lessThanOrEqualTo(concurrency)));
-        assertThat(response.totalShards, equalTo(shards));
-        assertThat(response.successfulShards, equalTo(shards));
-        assertThat(response.failedShards, equalTo(0));
-    }
-
-    public void testLimitConcurrentNodesForced() {
-        final int shards = 10;
-        var targetShards = new ArrayList<DataNodeRequestSender.TargetShard>(shards);
-        for (int i = 0; i < shards; i++) {
-            targetShards.add(
-                targetShard(
-                    new DataNodeRequest.Shard(new ShardId("index", "n/a", i), SplitShardCountSummary.UNSET),
-                    DiscoveryNodeUtils.builder("node-" + i).roles(Set.of(DATA_HOT_NODE_ROLE)).build()
-                )
-            );
-        }
+        // this will force requests to pile up
+        boolean forceConcurrent = randomBoolean();
 
         var concurrency = randomIntBetween(1, 2);
         AtomicInteger maxConcurrentRequests = new AtomicInteger(0);
@@ -381,11 +344,15 @@ public class DataNodeRequestSenderTests extends ComputeTestCase {
 
             sent.add(nodeRequest(node, shardIds));
             runWithDelay(() -> {
-                while (maxConcurrentRequests.get() < concurrency) {
-                    try {
-                        Thread.sleep(100);
-                    } catch (InterruptedException e) {
-                        throw new RuntimeException(e);
+                if (forceConcurrent) {
+                    int maxRetries = 0;
+                    // This forces the requests to pile up to test the concurrency limit is respected
+                    while (maxConcurrentRequests.get() < concurrency && maxRetries++ < 100) {
+                        try {
+                            Thread.sleep(100);
+                        } catch (InterruptedException e) {
+                            throw new RuntimeException(e);
+                        }
                     }
                 }
                 concurrentRequests.decrementAndGet();
@@ -393,7 +360,14 @@ public class DataNodeRequestSenderTests extends ComputeTestCase {
             });
         }));
         assertThat(sent.size(), equalTo(shards));
-        assertThat(maxConcurrentRequests.get(), equalTo(concurrency));
+        if (forceConcurrent) {
+            // Since we forced the concurrency, we should hit the exact concurrency limit here
+            assertThat(maxConcurrentRequests.get(), equalTo(concurrency));
+        } else {
+            // Since the requests are not forced to run concurrently, we might not hit the concurrency limit
+            // But we never have to exceed it
+            assertThat(maxConcurrentRequests.get(), is(lessThanOrEqualTo(concurrency)));
+        }
         assertThat(response.totalShards, equalTo(shards));
         assertThat(response.successfulShards, equalTo(shards));
         assertThat(response.failedShards, equalTo(0));

